Remove stale `variable_class_operations` set from `Target`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

**NOTE** This is a refactored PR since the cause of the errors was found to be a different reason.

I will leave the old description attached, the following description is only valid for commits after 0caa7d5e2375310b7a468b20a51bd66c7dd610cc

### Summary
Fixes #12953 

The following commits remove a stale attribute of the rust target called `variable_class_operations` which was used to keep track of all the variadics within the target but was not kept during the serialization step.

### Details and comments:

When performing serialization, we were forgetting to include `variable_class_operations` set of names in the state mapping. Since the nature of `TargetOperation` is to work as an enum of either `Instruction` instances or class aliases that would represent `Variadic` instructions. The usage of that structure was redundant, so it was removed. 

The following changes were made:
- `num_qubits` returns an instance of `u32`, callers will need to make sure they're dealing with a `NormalOperation`.
- `params` behaves more similarly, returning a slice of `Param` instances. Will panic if called on a `Variadic` operation.

A test case was added to check for something similar to what was mentioned by @doichanj in #12953.

### Old description
```
The following commit ensures that when the `num_qubits` or `params` are extracted from a variadic, we always return `None`.

### Details and comments
- The `num_qubits` property for `TargetOperation` returns an `Option<u32>` instead of panicking in the case that the instruction is a `Variadic`.
- The `params` property for `TargetOperation` returns an `Option<[&Param]>]` instance instead of panicking in the case of a `Variadic`.
```

raynelfss committed:
Fix: Return None for variadic properties.

raynelfss committed:
Merge branch 'main' into target-variadic

ElePT reviewed (commented):
Thanks @raynelfss! I think it would be good to also have a unit test for the fix, but looking at the issue it looks like you were not able to reproduce it locally, did you think of an alternative we could use?

raynelfss commented:
Thank you for checking this @ElePT, I couldn't replicate Jun's specific issue. But I do have an idea of how to replicate it. I will add unit tests soon. Thank you!

raynelfss committed:
Merge branch 'main' into target-variadic

raynelfss commented:
Actually the test-case for this has proven a bit harder to compose. This problem appeared to happen on previous versions of our code and only if Qiskit is ran with parallelization during some stage of `instruction_supported`. I will continue to look into this, the bug has been fixed but it would be great to include a test case for this as well.

raynelfss committed:
Merge branch 'main' into target-variadic

raynelfss committed:
Fix: Remove `variable_class_operations` from `Target`.
- When performing serialization, we were forgetting to include `variable_class_operations` set of names in the state mapping. Since the nature of `TargetOperation` is to work as an enum of either `Instruction` instances or class aliases that would represent `Variadic` instructiions. The usage of that structure was redundand, so it was removed.
- `num_qubits` returns an instance of `u32`, callers will need to make sure they're dealing with a `NormalOperation`.
- `params` behaves more similarly, returning a slice of `Param` instances. Will panic if called on a `Variadic` operation.
- Re-adapt the code to work without `variable_class_operations`.
- Add test case to check for something similar to what was mentioned by @doichanj in #12953.

mtreinish commented on test/python/transpiler/test_target.py:
```diff
@@ -1166,6 +1166,24 @@ def test_instruction_supported_no_args(self):
 
+    def test_target_serialization_preserve_variadic(self):
+        """Checks that variadics are still seen as variadic after serialization"""
+        from pickle import loads, dumps
```
You can move this import to the top of the module, we only need to do a runtime import if there is a potential cycle, but with pickle in the std lib that's not possible.

mtreinish commented on test/python/transpiler/test_target.py:
```diff
@@ -1166,6 +1166,24 @@ def test_instruction_supported_no_args(self):
+
+        target = Target("test", 1)
+        # Add variadic example gate with no properties.
+        target.add_instruction(XGate, None, "x_var")
```
It might be better to add a gate that doesn't have a fixed number of qubits, you can use `UnitaryGate` since it can have a variable number of qubits based on the size of the unitary matrix.

mtreinish reviewed (commented):
This LGTM it makes sense to remove the duplicated tracking indexmap if it's not needed. Just a couple of small inline comments about the tests.

raynelfss committed:
Fix: Use `UnitaryGate` as the example in test-case.
- Move import of `pickle` to top of the file.

mtreinish reviewed (approved):
LGTM, thanks for the quick update.
